### DISCUSSION TITLE
HOF File Output and Formatting Customization

### DISCUSSION THREAD

**Discussion by AaronRoseDA** [MilesCranmer/PySR#923] [Q&A]
Greetings,

Is there a way to customize the behavior and format of the HOF output, including what’s saved in the output directory? For example, the model might progress through complexities of 4, 6, 8, and 12 with decreasing loss. But if a better expression is found at complexity 9, the 12-complexity expression gets overwritten. Is there a way to save every expression that improves upon the previous one (maybe only if it improves the loss from the previous best expression with a higher complexity)? From a research standpoint, it would be valuable to observe how the training evolves in more detail. This information might already be available in the model object after calling fit, but I'm not sure. Ideally, having access to metrics like Complexity, Loss, Score, Equation, Runtime, Min Error, Max Error, etc., would be huge!

![image](https://github.com/user-attachments/assets/5c2a008e-44d5-40db-a1bb-0339c0fbf921)

```
model = PySRRegressor(
    model_selection="best",
    run_id=f'SkewNormal - {datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S.%f")[:-4]}',
    niterations=10000000,
    binary_operators=["+", "-", "*", "/"],
    unary_operators=["log", "sqrt", "square","exp"
                     # , "sin", "cos", "tan"
                     ],
    maxsize=12,
    warm_start=True,
    temp_equation_file=False,
    output_directory="C:/Users/person/Desktop/output/Skewnormal",
    verbosity=1,
    timeout_in_seconds=120,
    # early_stop_condition="f(loss, complexity) = (loss <= 5.526e-15) && (complexity < 6)",
    print_precision=6
)
```

Thank you!

**Comment by MilesCranmer** [comment]
Yes! You can use TensorBoardLoggerSpec: https://ai.damtp.cam.ac.uk/pysr/examples/#12-using-tensorboard-for-logging, which logs the full history of the Pareto front to disk in tensorboard format. This doubles as a way to interface with a GUI and hyperparameter sweep tools.
